lib/spack/spack/build_environment.py

```diff
@@ -1000,7 +1000,7 @@ def __init__(self, *specs: spack.spec.Spec, context: Context) -> None:
 
     def set_all_package_py_globals(self):
         """Set the globals in modules of package.py files."""
-        for dspec, flag in chain(self.external, self.nonexternal):
+        for dspec, flag in chain(self.nonexternal, self.external):
```

Comment:
Alternative summary:

`msmpi`'s package.py does:

```
    def setup_dependent_package(self, module, dependent_spec):
        ...
        self.spec.mpicc = dependent_module.spack_cc
```

Externals are (almost always) leaves, so this PR ensures that `set_package_py_globals` runs on `hdf5` (the parent) before `setup_dependent_package` runs on `msmpi` (the child).

If it's valid in any circumstance for a child's `setup_dependent_package` function to access attributes of the parent (I think it is), then I think a change (like this PR) is needed.

Another way to approach this is to split the loop into two loops: one for `setup_dependent_package`, and one for `set_package_py_globals`; if it did, the externals-before-non-externals order could be preserved (at least for `msmpi`).

No unit tests fail with the rearrangement in this PR, and I don't see a reason to prefer the original order from looking through `set_package_py_globals`, so I'm curious if @haampie has a preference (based on `git blame`, the latest person to edit this section).

Reply:
Looking at the code, I think this is also wrong. We should probably split the loop in two, and:
1. First run `set_package_py_globals` on everything
2. Then re-run the loop and call `setup_dependent_package`.

Also, wouldn't it make sense to run the loop as a topo-sort from leaves to root (instead of external vs. non-external)? @haampie